sdk/storage/src/common/crypt.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "common/crypt.hpp"
+
+#include "common/storage_common.hpp"
+
+#ifdef _WIN32
+#include <Windows.h>
+#include <bcrypt.h>
+#else
+#include <openssl/bio.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#endif
+
+#include <stdexcept>
+
+namespace Azure { namespace Storage {
+
+#ifdef _WIN32
+  std::string HMAC_SHA256(const std::string& text, const std::string& key)
+  {
+    struct AlgorithmProviderInstance
+    {
+      BCRYPT_ALG_HANDLE Handle;
+      std::size_t ContextSize;
+      std::size_t HashLength;
+
+      AlgorithmProviderInstance()
+      {
+        NTSTATUS status = BCryptOpenAlgorithmProvider(
+            &Handle, BCRYPT_SHA256_ALGORITHM, NULL, BCRYPT_ALG_HANDLE_HMAC_FLAG);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("failed to call bcrypt function");
+        }
+        DWORD objectLength = 0;
+        DWORD dataLength = 0;
+        status = BCryptGetProperty(
+            Handle,
+            BCRYPT_OBJECT_LENGTH,
+            (PBYTE)&objectLength,
+            sizeof(objectLength),
+            &dataLength,
+            0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("failed to call bcrypt function");
+        }
+        ContextSize = objectLength;
+        DWORD hashLength = 0;
+        status = BCryptGetProperty(
+            Handle, BCRYPT_HASH_LENGTH, (PBYTE)&hashLength, sizeof(hashLength), &dataLength, 0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("failed to call bcrypt function");
+        }
+        HashLength = hashLength;
+      }
+
+      ~AlgorithmProviderInstance() { BCryptCloseAlgorithmProvider(Handle, 0); }
+    };
+
+    static AlgorithmProviderInstance AlgorithmProvider;
+
+    std::string context;
+    context.resize(AlgorithmProvider.ContextSize);
+
+    BCRYPT_HASH_HANDLE hashHandle;
+    NTSTATUS status = BCryptCreateHash(
+        AlgorithmProvider.Handle,
+        &hashHandle,
+        (PUCHAR)context.data(),
+        (ULONG)context.size(),
+        (PUCHAR)key.data(),
+        (ULONG)key.length(),
+        0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("failed to call bcrypt function");
```

Comment:
>throw std::runtime_error("failed to call bcrypt function"); [](start = 6, length = 59)

This exception is too general. I think each failing has it's own reason. Although it's not high priority here but please add an issue for us to revisit that we need more detailed exception message here.

Reply:
fixed